### PR TITLE
Support XML config for fastdds-discovery-server [13783]

### PR DESCRIPTION
**Pull request by VictorMMaroto**
Documentation PR: eProsima/Fast-DDS-docs#351

Signed-off-by: Victor M Maroto <victormaroto@eprosima.com>

### CODE REVIEW

**Comment by JLBuenoLopez-eProsima** [comment]
CI failures are unrelated

File: tools/fds/server.cpp [hunk 1]
```diff
@@ -99,15 +104,79 @@ int fastdds_discovery_server(
...
+        if (!(participantQos.wire_protocol().builtin.discovery_config.discoveryProtocol ==
+                eprosima::fastrtps::rtps::DiscoveryProtocol::SERVER ||
+                participantQos.wire_protocol().builtin.discovery_config.discoveryProtocol ==
+                eprosima::fastrtps::rtps::DiscoveryProtocol::BACKUP) ||
+                participantQos.wire_protocol().prefix == prefix_cero)
+        {
+            std::cout << "The provided configuration is not valid. Participant must be either SERVER or BACKUP. " <<
+                "Server id is mandatory if not defined in the XML file: use -i or --server-id option." << std::endl;
+            return 1;
+        }
```
Comment (MiguelCompany) [inline tools/fds/server.cpp:179]: I would have different checks for the discovery protocol and the guid prefix, so the error message is specific.

File: tools/fds/server.cpp [hunk 2]
```diff
@@ -99,15 +104,79 @@ int fastdds_discovery_server(
...
 
+    if (nullptr != options[XML_FILE])
+    {
+        sXMLConfigFile = options[XML_FILE].arg;
+        if (sXMLConfigFile.length() > 0)
+        {
+            size_t delimiter_pos = sXMLConfigFile.find(delimiter);
+            if (std::string::npos != delimiter_pos)
+            {
+                profile = sXMLConfigFile.substr(delimiter_pos + 1, sXMLConfigFile.length());
```
Comment (MiguelCompany) [inline tools/fds/server.cpp:115]: If we change the profile position, this should be changed

File: tools/fds/server.h [hunk 3]
```diff
@@ -97,7 +105,15 @@ const option::Descriptor usage[] = {
...
-      "\t$ " FAST_SERVER_BINARY " -i 0 -l localhost -p 14520"},
+      "\t$ " FAST_SERVER_BINARY " -i 0 -l localhost -p 14520\n\n"
+
+      "\t6. Launch a server with id 0 (first on ROS_DISCOVERY_SERVER) reading\n"
+      "\t   default configuration from XML file.\n\n"
+      "\t$ " FAST_SERVER_BINARY " -i 0 -x config.xml\n\n"
+
+      "\t6. Launch a server with id 0 (first on ROS_DISCOVERY_SERVER) reading\n"
+      "\t   specific profile_name configuration from XML file.\n\n"
+      "\t$ " FAST_SERVER_BINARY " -i 0 -x config.xml@profile_name"},
```
Comment (MiguelCompany) [inline tools/fds/server.h:116]: If we change the profile position, this should be changed

File: tools/fds/server.cpp [hunk 4]
```diff
@@ -168,65 +244,70 @@ int fastdds_discovery_server(
 
     // Retrieve first IP address
     pOp = options[IPADDRESS];
-    if (nullptr == pOp)
+    if (nullptr == pOp && (nullptr == options[XML_FILE] || nullptr != pO_port))
```
Comment (MiguelCompany) [inline tools/fds/server.cpp:247]: I am not sure why the last comparison is `!=`. Would you mind explaining?

File: test/system/tools/fds/tests.py [hunk 5]
```diff
@@ -107,24 +101,424 @@ def test_fast_discovery_closure(fast_discovery_tool):
...
+
+def test_fast_discovery_closure(fast_discovery_tool):
+    """Test that discovery command closes correctly."""
+    command = [fast_discovery_tool, '-i', '0']
+
+    output, err, exit_code = send_command(command)
+
+    EXPECTED_CLOSURE = "### Server shut down ###"
+
+    check_output(output, err, EXPECTED_CLOSURE, False)
```
Comment (MiguelCompany) [inline test/system/tools/fds/tests.py:156]:
```suggestion
    exit_code = check_output(output, err, EXPECTED_CLOSURE, False)
```

File: test/system/tools/fds/tests.py [hunk 6]
```diff
@@ -107,24 +101,424 @@ def test_fast_discovery_closure(fast_discovery_tool):
...
+    EXPECTED_SERVER_ADDRESS = []
+    udpv4 = uri_profile.getElementsByTagName('udpv4')
+    for elem in udpv4:
+        address2 = elem.getElementsByTagName('address')[0].firstChild.data
+        port2 = elem.getElementsByTagName('port')[0].firstChild.data
+        if port2[0] == '0':
+            port2 = port2[1:]
+        EXPECTED_SERVER_ADDRESS.append("UDPv4:[" + address2 + "]:" + port2)
+
+    command = [fast_discovery_tool, '-x', XML_file_path + '@UDP_server_two']
```
Comment (MiguelCompany) [inline test/system/tools/fds/tests.py:216]: If we change the profile position, this should be changed. Same elsewhere.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Apart from my comments below, it is shocking for me that the profile name is specified after the `@`. On HTTP URIs, the username is specified with `http://username@host/`

File: test/system/tools/fds/tests.py (hunk 6 above)
Comment (JLBuenoLopez-eProsima) [reply test/system/tools/fds/tests.py]: Done in 6dd1ad92f

File: test/system/tools/fds/tests.py (hunk 5 above)
Comment (JLBuenoLopez-eProsima) [reply test/system/tools/fds/tests.py]: Done in 26047a6a9

File: tools/fds/server.cpp (hunk 4 above)
Comment (JLBuenoLopez-eProsima) [reply tools/fds/server.cpp:280]: Done in 7d6428651

File: tools/fds/server.h (hunk 3 above)
Comment (JLBuenoLopez-eProsima) [reply tools/fds/server.h]: Done in 6dd1ad92f

File: tools/fds/server.cpp (hunk 2 above)
Comment (JLBuenoLopez-eProsima) [reply tools/fds/server.cpp]: Done in 6dd1ad92f

File: tools/fds/server.cpp (hunk 1 above)
Comment (JLBuenoLopez-eProsima) [reply tools/fds/server.cpp]: Done in f4bbf6a71

File: test/system/tools/fds/test_wrong_xml_discovery_server.xml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                    <locator>
+                        <udpv4>
+                            <address>127.0.0.5</address>
+                            <port>11817</port>
+                        </udpv4>
+                    </locator>
+                </metatrafficUnicastLocatorList>
+            </builtin>
+        </rtps>
+    </participant>
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/system/tools/fds/test_wrong_xml_discovery_server.xml]: Done in e155ce1d0

File: tools/fds/server.cpp [hunk 8]
```diff
@@ -70,7 +68,9 @@ int fastdds_discovery_server(
     std::vector<option::Option> options(stats.options_max);
     std::vector<option::Option> buffer(stats.buffer_max);
     option::Parser parse(usage, argc, argv, &options[0], &buffer[0]);
-    std::string sXMLConfigFile = "", delimiter = "@", profile = "";
+    std::string sXMLConfigFile = "";
+    std::string delimiter = "@";
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply tools/fds/server.cpp]: Done in 7d6428651

File: tools/fds/server.cpp [hunk 9]
```diff
@@ -167,14 +173,18 @@ int fastdds_discovery_server(
     if (nullptr == pOp)
     {
         fastrtps::rtps::GuidPrefix_t prefix_cero;
-        if (!(participantQos.wire_protocol().builtin.discovery_config.discoveryProtocol ==
+        if (participantQos.wire_protocol().prefix == prefix_cero)
+        {
+            std::cout << "Server id is mandatory if not defined in the XML file: use -i or --server-id option." <<
+                std::endl;
```
Comment (MiguelCompany) [inline tools/fds/server.cpp:179]:
```suggestion
                std::endl;
            return 1;
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Just one change and we are good to go.

File: tools/fds/server.cpp (hunk 9 above)
Comment (JLBuenoLopez-eProsima) [reply tools/fds/server.cpp:179]: :facepalm:

**Review by MiguelCompany — APPROVED** [review]
LGTM

**Comment by MiguelCompany** [comment]
@JLBuenoLopez-eProsima Two of the new tests failed on aarch64

**Comment by MiguelCompany** [comment]
@richiprosima Please test aarch64

**Pull request merged**